index.bs

```diff
@@ -209,83 +251,97 @@ enum OperandType {
   sequence<long> dimensions;
 };
 </script>
 
-## Operand ## {#api-operand}
+## MLOperand ## {#api-mloperand}
 <script type=idl>
 [SecureContext, Exposed=Window]
-interface Operand {};
+interface MLOperand {};
 </script>
 
-## NeuralNetworkContext ## {#api-neuralnetworkcontext}
+## MLModelBuilder ## {#api-mlmodelbuilder}
 
-The {{NeuralNetworkContext}} interface represents a global state of neural network compute workload and execution processes.
+The {{MLModelBuilder}} interface defines a set of operations as identified by the [[#usecases]] that can be composed into a computational graph. It also represents the intermediate state of a graph building session.
 
 <script type=idl>
+// WebGL specification:
+// (https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.4)
 [SecureContext, Exposed=Window]
-interface NeuralNetworkContext {
-  ModelBuilder createModelBuilder();
+interface WebGLBuffer {
+  // ...
 };
-</script>
 
-## ModelBuilder ## {#api-modelbuilder}
-
-The {{ModelBuilder}} interface defines a set of operations as identified by the [[#usecases]] that can be composed into a computational graph. It also represents the intermediate state of a graph building session.
+// WebGPU specification:
+// (https://gpuweb.github.io/gpuweb/#buffer-interface)
+[SecureContext, Exposed=Window]
+interface GPUBuffer {
+  // ...
+};
 
-<script type=idl>
-typedef record<DOMString, Operand> NamedOperands;
+typedef record<DOMString, MLOperand> MLNamedOperands;
 
 [SecureContext, Exposed=Window]
-interface ModelBuilder {
-  // Create an operand that represents a model input.
-  Operand input(DOMString name, OperandDescriptor desc);
+interface MLModelBuilder {
+  // Construct the model builder from the context.
+  constructor(MLContext context);
+
+  // Create an operand for a model input.
+  MLOperand input(DOMString name, MLOperandDescriptor desc);
+
+  // Create an operand for a model constant.
+  MLOperand constant(MLOperandDescriptor desc, ArrayBufferView bufferView);
 
-  // Create an operand that represents a model constant.
-  Operand constant(OperandDescriptor desc, ArrayBufferView value);
+  // Create an operand for a model constant from a WebGL buffer.
+  MLOperand constant(MLOperandDescriptor desc, WebGLBuffer glBuffer, 
+                     optional long offset = 0, optional long size);
+
+  // Create an operand for a model constant from a WebGPU buffer.
+  MLOperand constant(MLOperandDescriptor desc, GPUBuffer gpuBuffer, 
+                     optional long offset = 0, optional long size);
 
   // Create a single-value operand from the specified number of the specified type.
-  Operand constant(double value, optional OperandType type = "float32");
+  MLOperand constant(double value, optional MLOperandType type = "float32");
 
   // Create a model that encapsulates the composition of operands by identifying the output operands.
-  Model createModel(NamedOperands outputs);
+  MLModel createModel(MLNamedOperands outputs);
```

Comment:
Should we still keep `MLModel` building device agnostic and leave the device association to compile time? Say to support `model.compile(glContext)` and `model.compile(gpuDevice)`. With that, developer can build the same model and compile for different devices.

Reply:
There are interop use cases where the model constants (e.g. weights, etc.) are already uploaded to the GPU buffers before the model is created. In that case it will be inefficient to require the caller to make a roundtrip of that data back to the CPU memory just to construct a device-agnostic model.

As we position WebNN also as a backend API for frameworks, the notion of device-agnostic graph could be in a way of the interop efficiency since by the time the backend graph is initiated (by the frameworks), the model constants may already be mapped to the memory hosted by the frameworks, and for the frameworks already dealing with device resources, those constants could already be uploaded to the device.

Reply:
It makes sense, thanks for the explanation. Given the `MLModel` is now device specific, would we consider merging it into the `MLCompilation`? Let's say replace the `MLModel createModel(MLNamedOperands outputs)` with `Promise<MLCompilation> compile(MLNamedOperands outputs)`.

Reply:
@wchao1115 , any thoughts about this idea?

Reply:
I can see the logic to it. I think it makes sense.

Reply:
I think it would also help the use case of [operation-specific API](https://github.com/webmachinelearning/proposals/issues/2#issuecomment-773441501). For example, the `webml_create_conv2d` could be just compiling a `conv2d`. e.g.
```js
const conv2d_op = await builder.compile({output: builder.conv2d(input, filter)});
```
 For fused conv2d, it could be compiling `relu`, `add` and `conv2d`. e.g.
```js
const fused_conv2d_op = await builder.compile({output: builder.relu(builder.add(builder.conv2d(input, filter), bias))});
```

@pyu10055

Reply:
Yes, that looks much more compact. I've made the suggested changes, please take a look. 

However, since `MLModel` is removed, calling the class `MLModelBuilder` may no longer be appropriate. We could rename `MLModelBuilder` to something more relevant, or rename `MLCompilation` to `MLModel` (as in *compiled model*)

Reply:
I see your point that device specific model can improve overall performance, there are cases the data is on CPU but the user want to execute it on GPU, does this API requires framework to upload the data to GPU as Texture and call WebNN API?
Will it be more efficient for WebNN to copy CPU data directly to its backend data structure without going through the Texture intermediate layer?
If it is really tied to the Web APIs (WebGL, WebGPU, WASM), an op level API would make much more sense, since most ML frameworks already has a graph building APIs. And the WebNN API can be stateless and functional API.

Reply:
No, it doesn't require the framework to upload. The constants, inputs, and outputs can still be in `ArrayBufferView` in which case the implementation of webnn should still upload the resources themselves if it is running on a hardware device. This change just adds an additional capacity for the API to deal with uploaded resources.

Reply:
> However, since `MLModel` is removed, calling the class `MLModelBuilder` may no longer be appropriate. We could rename `MLModelBuilder` to something more relevant, or rename `MLCompilation` to `MLModel` (as in _compiled model_)

Agree. I am fine with renaming `MLCompilation` to `MLModel`. 

Actually, in this spec, a _model_ means a _graph_ and we use _graph_ everywhere. To avoid this translation, probably we could rename 
 * `MLModelBuilder` -> `MLGraphBuilder`
 * `MLModelBuilder.compile` -> `MLGraphBuilder.build`
 * `MLCompilation` -> `MLGraph`

e.g.
```js
const graph = await builder.build({c: builder.matmul(a, b)});
const outputs = await graph.compute(inputs);
```

Reply:
> If it is really tied to the Web APIs (WebGL, WebGPU, WASM), an op level API would make much more sense, since most ML frameworks already has a graph building APIs. And the WebNN API can be stateless and functional API.

@pyu10055 , the `compute` API of WebNN is stateless. The graph can contain just a single op (e.g. `conv2d`)  or a few decomposed ops that can be fused by native ML APIs (e.g. `conv2d` + `add` + `relu`). With this PR, now the `compute` can accept input and output resources on GPU device without copying between GPU and CPU. 

Does this meet the op level execution requirement of the ML frameworks mentioned in https://github.com/webmachinelearning/proposals/issues/2#issuecomment-773441501?

Reply:
Additionally, to support the eager execution of frameworks (e.g. TF.js) efficiently, we may consider the point made by @wchao1115 in the comment https://github.com/webmachinelearning/proposals/issues/2#issuecomment-781057233 that

> 2\. When an operation is invoked from a framework to the underlying platform, the data type and layout format must be in a form that can be readily usable by the operation without incurring additional copies or conversions. 

The framework may compute multiple ops without accessing the intermediate results. So if we allow passing the output of a compute to the next compute without making the data available, it would help the underlying implementation hide the additional copies and conversions. The idea is illustrated by following pseudo framework integration code:

```js
// assume conv2d_op, add_op, relu_op are compiled single-op graphs
// input = tf.tensor(values);
// conv2d = tf.conv2d(input, filter);
conv2d.mlOutput = conv2d_op.compute({'input': {resource: input.arrayBuffer}).output; // webnn backend
// add = tf.add(conv2d, bias);
add.mlOutput = add_op.compute({'a': conv2d.mlOutput}).output; // webnn backend
// relu = tf.relu(add);
relu.mlOutput = relu_op.compute({'input': add.mlOutput}).output;  // webnn backend
// await relu.data();
await relu.mlOutput.readDataAsync({resource: relu.arrayBuffer});  // webnn backend
```

It may require the following changes:
 * change the `MLOutput` to an interface that can read the output data (`MLOutputData`) back asynchronously
 * `compute` can accept `MLInput` or `MLOutput` in `MLNamedInputs`
 * Because `MLOutput` can read output data back asynchronously, the `compute` method can be synchronous. 

```js
dictionary MLInput {
  required (MLBuffer or WebGLTexture or GPUTexture) resource;
  sequence<long> dimensions;
};
dictionary MLOutputData {
  (MLBuffer or WebGLTexture or GPUTexture) resource;
  sequence<long> dimensions;
};
interface MLOutput {
  Promise<MLOutputData> readDataAsync(optional MLOutputData data); 
};
typedef record<DOMString, (MLInput or MLOutput)> MLNamedInputs;
typedef record<DOMString, MLOutput> MLNamedOutputs;
[SecureContext, Exposed=Window]
interface MLCompilation {
  MLNamedOutputs compute(MLNamedInputs inputs, optional sequence<DOMString> requiredOutputs);
};
```

Thoughts?

Reply:
We should discuss eager execution in a separate issue. Please feel free to post an issue for that. 

I think in general eager execution is there to aid model development by making the computed result immediately available. I feel it's an issue separated from how to make the operation running as efficiently as possible because the former is for debugging while the latter is for perf. 

The general steps of operation execution is build > compile > dispatch > execute. The compile step is where activities like constant folding and fusions occur. Fusion in particular is the result of fusing multiple semantic operations that were previously built as a subgraph into an atomic executable unit. Additionally for compute platforms like the GPU, the compile step is the time that JITing of shader fragments and preprocessing of constant weight data (aka. weight-packing) can occur.

Once the subgraph is compiled down, the compiled operations can be dispatched or placed into the command buffers. The placement of dispatches can be done by multiple CPU threads simultaneously while the GPU is executing the command buffers that were previously dispatched, preventing stalls and bubbles while appropriately saturating the GPU. 

Under the hood, eager execution is all these 4 steps sequentially executed in one go.

Reply:
Thanks for the explanation @wchao1115 !

What I propose here is to support the op-level execution of a framework, like TensorFlow.js, based on current graph API. The framework may rely on the op-level execution capability to enable the eager mode. I am not proposing to add the eager execution capability into WebNN API itself. I agree the later one should be discussed in a separate issue.

As @pyu10055 mentioned in above https://github.com/webmachinelearning/webnn/pull/149#discussion_r591039364, the framework can execute ops by computing single-op graph or fused-ops graph of WebNN. However, current `compute` method couples the graph execution and output data reading-back/conversion together. This design would not be optimal when frameworks execute multiple ops (multiple WebNN graphs) without accessing the intermediate results. Some frameworks, like TensorFlow.js, infer a model by executing all its ops but only access the output data of the final ops. If using today's `compute` API, it would convert or read back output data after each op's compute that is not necessary. So I propose to decouple the graph execution and output data reading-back into two steps. `compute` just returns an opaque `MLOutput` object. The `MLOutput` can be passed to the next `compute` as input that doesn't cause data reading-back or conversion. Framework only invokes `readDataAsync` when need accessing the output data.

Does it make sense?

Reply:
This is a good issue. I'll work on this after completing this PR. 

I understand the scenario and I certainly agree that `compute` should become a sync all. But I think the fact that `MLOutput` now specifies the union of different output resource types including the hardware resources, and that the validity of the use case of those resources are already dependent upon the same input resource type i.e. a `WebGLBuffer` output resource is only available when the input is also of `WebGLBuffer` type, should already mean that `compute` will produce its result on a corresponding resource type. 

Basically if the concern is over the repeated downloads cost of intermediate results across multiple disconnected single operations, the answer is if your input resource is a GPU buffer, the output is going to have to be in a GPU buffer, and not something that is automatically downloaded to an array buffer. This is true for whether the execution is over just 1 operator or an entire graph i.e. if you want to download the result, you will need to download from the output resource type the execution has placed its result into. If all the executions is happening on an array buffer, then obviously no download would be needed, and all that lies outside of the WebNN API.

Reply:
My concern is about if the input and output are array buffers but the compute device is a gpu or an accelerator, the implementation needs to upload and download the data repeatedly. 

Another aspect is about memory relayout, even for cpu, some native APIs, like oneDNN, use [optimized memory layout](https://oneapi-src.github.io/oneDNN/dev_guide_understanding_memory_formats.html) for hardware acceleration. Since the memory layout/relayout brings significant overhead, so it would be good to avoid that if the intermediate results are not accessed. I am not sure whether gpu device has similar memory relayout overhead. Any insights @wchao1115 ?

Reply:
> My concern is about if the input and output are array buffers but the compute device is a gpu or an accelerator, the implementation needs to upload and download the data repeatedly.

When the call creates an `MLContext` they have an option to either create a default context with preference options e.g. power preference, etc. -or- create a context from an existing device. This distinction is important because if a default context is created, it is implied that the API implementation chooses and completely owns the device it uses for subsequent execution. In this use case, the caller has no way to know if the underlying device is a HW resource or a CPU memory because -- it is up to the API to decide. So, the output must be in a form that can readily be used, which is an array buffer view. If the underlying device happens to be a GPU, the output is automatically downloaded. This contract is important because it is how the API is used in a device-agnostic way -- from the caller's point of view. 

But if the context is explicitly created by an existing device, it means the device is under the control of the caller, which implies that the caller is the one who would manage the lifetime of the resources created by that device. In this use case, the output will be on the device's resource, and that the caller would own download the data from the resource at the right time in their scenario. This use case is what being described as the one that needs the atomic operations (aka the op-level execution) because its when the caller would have the control on the device resource including data uploads/downloads, layout assignment, and block layout processing etc.

So, in short, yes, for the default context, the data is automatically uploaded/downloaded if the underlying device needs them, but that is not the scenario that would call for the efficient op-level execution where the explicit resource management is a big implicit part of the said "efficiency" to be had.

Reply:
My concern is about the efficient op-level execution on CPU. This scenario was mentioned by @pyu10055 in https://github.com/webmachinelearning/proposals/issues/2#issuecomment-773441501 for framework's CPU - WebAssembly backend. 

The CPU can only be selected for default context. However as the default context would automatically upload/download data, it may bring unnecessary memory re-layout overhead as I mentioned. So my proposal is to allow the user code to control when to relayout/download data to array buffer for default context. I suppose this won't change the device-agnostic view of the default context. And it would also enable the efficient op-level execution for framework's CPU - WebAssembly scenario.

Reply:
I agree with @huningxin that, the returned tensors should be tensor handles that can be downloaded by a separate API.

Reply:
@huningxin and @pyu10055, on the topic of wasm interop, I had a chance to chat with people who are more familiar with wasm, and they pointed me to this [`WebAssembly.Memory`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory) object, which allows for an efficient interop with JS `ArrayBuffer`. In short, the memory passed between wasm and JS is always part of a wasm `Memory` object that can be the backing store for a JS ArrayBuffer. Also, take a look at a more elaborate sample on JS/WASM interop [here](https://rob-blackbourn.github.io/blog/webassembly/wasm/array/arrays/javascript/c/2020/06/07/wasm-arrays.html)

Thru the wasm memory object, a wasm library can always interop with a JS ArrayBuffer with no copy required, and likewise in the opposite direction -- an ArrayBuffer can also be created from a wasm memory and used as an output buffer of a JS API call. Hope this helps clarify this topic.

Reply:
Thanks for the pointers @wchao1115 ! It really helps.

With the `ArrayBufferView` support by `MLGraphBuilder.constant` and `MLGraph.compute`, a wasm library can interop with WebNN API with `ArrayBufferView` backed by the `WebAssembly.Memory.buffer`. There are no issues.

However, as I mentioned in https://github.com/webmachinelearning/webnn/issues/156, there would be other two requirements for an efficient wasm interops:
1. a wasm library could create a `MLContext` from the CPU device, otherwise there might be across device data copies that a wasm library doesn't want. 
2. for op-level execution (by single-op `MLGraph`), a wasm library could control when the output data is available, otherwise there might be tensor memory re-layout overhead between native ML API and WebNN for each op's execution that are not always needed.

Reply:
> adding a device preference option in the `MLContextOptions`. This option will be complementary to the existing power preference.

I like it!

> what prevents the caller from compiling all the 3 `conv2d` ops into a single compiled graph before executing it in one go?

That's because the framework provides the op API. When executing a graph, the user code just calls the op API to execute the operations of the graph one-by-one. The framework doesn't have the knowledge of the graph, it just executes ops.

For this scenario, the framework doesn't know the user code will execute 3 `conv2d`. When user code calls a `conv2d`, the framework just computes that `conv2d`. The framework may use a platform-specific tensor layout in the `conv2d` implementation. Until the user code calls `h.data()`, the framework starts to convert and copy the tensor data into a `ArrayBufferView` in plain layout. 

So if the framework uses WebNN to compute `conv2d`, it would hurt the performance if WebNN always converts and copies the output tensor data to `ArrayBufferView` for each compute.

> The point is that the end result of executing a publicly defined API must not be platform-specific

I agree with you, at the interops point, it should use the plain/standard tensor layout.

Reply:
> That's because the framework provides the op API. When executing a graph, the user code just calls the op API to execute the operations of the graph one-by-one. 

That doesn't sound like what a framework does. Are you saying that there are frameworks that never compiles a graph internally and just relying on individual op execution?

Reply:
Framework may provide both op API and model API, for example TensorFlow.js provides [Operations API](https://js.tensorflow.org/api/latest/#Operations) and [Models/Loading API](https://js.tensorflow.org/api/latest/#Models-Loading). It would depend on what API the user code uses. If the user code uses the op API to execute a graph, the framework won't get the graph but individual ops.

Actually, webnn-polyfill uses TensorFlow.js op API. I suppose it is a common scenario. cc @pyu10055 .

Reply:
@wchao1115 @huningxin Thank you Ningxin for brought up this point, op level execution is common for eager execution, especially in JS environment, users could creating graph on the fly. There will be no specific graph boundary, framework would not be able to know which are the intermediate tensors that users would not need to access.